File: app/src/main/java/com/sjani/usnationalparkguide/UI/Details/DetailViewModel.java
package com.sjani.usnationalparkguide.UI.Details;


import com.sjani.usnationalparkguide.Data.AlertEntity;
import com.sjani.usnationalparkguide.Data.CampEntity;
import com.sjani.usnationalparkguide.Data.FavParkEntity;
import com.sjani.usnationalparkguide.Data.ParkEntity;
import com.sjani.usnationalparkguide.Data.TrailEntity;
import com.sjani.usnationalparkguide.Utils.ParkRepository;

import java.util.List;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MediatorLiveData;
import androidx.lifecycle.ViewModel;

public class DetailViewModel extends ViewModel {

    private static final String TAG = DetailViewModel.class.getSimpleName();
    private final LiveData<ParkEntity> park;
    private final ParkRepository repository;
    private LiveData<TrailEntity> trail;
    private LiveData<CampEntity> camp;
    private LiveData<FavParkEntity> favPark;
    private LiveData<List<FavParkEntity>> favParks;
    private LiveData<List<TrailEntity>> trails;
    private LiveData<List<CampEntity>> camps;
    private LiveData<List<AlertEntity>> alerts;
    private MediatorLiveData<List<FavParkEntity>> mediatorFavPark = new MediatorLiveData<>();
    private String parkCode;
    private String apiKey;
    private String trailApiKey;
    private String latLong;
    private String fields;

    public DetailViewModel(ParkRepository repository, String parkCode, String apiKey, String trailApiKey, String latLong, String fields) {
        this.repository = repository;
        this.parkCode = parkCode;
        this.apiKey = apiKey;
        this.trailApiKey = trailApiKey;
        this.latLong = latLong;
        this.fields = fields;
        park = this.repository.getParkFromDb(parkCode);
    }

    public LiveData<ParkEntity> getPark() {
        return park;
    }

    public void setFavPark(FavParkEntity park) {
        this.repository.updateFav(park);
    }

    public LiveData<List<FavParkEntity>> getfavpark() {
        favParks = this.repository.getFavPark();
        return favParks;
    }

    public void clearFav() {
        this.repository.clearFav();
    }


    public LiveData<List<TrailEntity>> getTrails() {
        trails = this.repository.getTrailListFromDb(trailApiKey, latLong);
        return trails;
    }

    public LiveData<List<CampEntity>> getCamps() {
        camps = this.repository.getCampListFromDb(parkCode, fields, apiKey);
        return camps;
    }

    public LiveData<List<AlertEntity>> getAlerts() {
        alerts = this.repository.getAlertListFromDb(parkCode, apiKey);
        return alerts;
    }
}

File: app/src/main/java/com/sjani/usnationalparkguide/ui/viewmodel/DetailViewModel.kt
package com.sjani.usnationalparkguide.ui.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewModelScope
import com.sjani.usnationalparkguide.data.entity.AlertEntity
import com.sjani.usnationalparkguide.data.entity.CampEntity
import com.sjani.usnationalparkguide.data.entity.FavParkEntity
import com.sjani.usnationalparkguide.data.entity.ParkEntity
import com.sjani.usnationalparkguide.data.entity.TrailEntity
import com.sjani.usnationalparkguide.data.model.CurrentWeather
import com.sjani.usnationalparkguide.data.repository.ParkRepository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch

data class DetailUiState(
    val isLoading: Boolean = true,
    val park: ParkEntity? = null,
    val isFavorite: Boolean = false,
    val weather: CurrentWeather? = null,
    val trails: List<TrailEntity> = emptyList(),
    val campgrounds: List<CampEntity> = emptyList(),
    val alerts: List<AlertEntity> = emptyList(),
    val error: String? = null
)

class DetailViewModel(
    private val repository: ParkRepository,
    private val parkCode: String,
    private val latLong: String,
    private val npsApiKey: String,
    private val trailApiKey: String,
    private val weatherApiKey: String,
    private val campgroundFields: String
) : ViewModel() {
    
    private val _uiState = MutableStateFlow(DetailUiState())
    val uiState: StateFlow<DetailUiState> = _uiState.asStateFlow()
    
    val park = repository.getPark(parkCode)
        .stateIn(viewModelScope, SharingStarted.Lazily, null)
    
    val trails = repository.getAllTrails()
        .stateIn(viewModelScope, SharingStarted.Lazily, emptyList())
    
    val campgrounds = repository.getAllCampgrounds()
        .stateIn(viewModelScope, SharingStarted.Lazily, emptyList())
    
    val alerts = repository.getAllAlerts()
        .stateIn(viewModelScope, SharingStarted.Lazily, emptyList())
    
    val favorites = repository.getAllFavorites()
        .stateIn(viewModelScope, SharingStarted.Lazily, emptyList())
    
    init {
        loadData()
    }
    
    private fun loadData() {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true)
            try {
                // Load trails
                repository.refreshTrails(trailApiKey, latLong)
                
                // Load campgrounds
                repository.refreshCampgrounds(parkCode, campgroundFields, npsApiKey)
                
                // Load alerts
                repository.refreshAlerts(parkCode, npsApiKey)
                
                // Load weather
                val weather = repository.getCurrentWeather(latLong, weatherApiKey)
                _uiState.value = _uiState.value.copy(
                    isLoading = false,
                    weather = weather,
                    error = null
                )
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(
                    isLoading = false,
                    error = e.message ?: "Unknown error"
                )
            }
        }
    }
    
    fun addToFavorites(park: ParkEntity) {
        viewModelScope.launch {
            val favParkEntity = FavParkEntity.fromParkEntity(park)
            repository.addFavorite(favParkEntity)
        }
    }
    
    fun removeFromFavorites(parkId: String) {
        viewModelScope.launch {
            repository.removeFavorite(parkId)
        }
    }
    
    fun refreshWeather() {
        viewModelScope.launch {
            val weather = repository.getCurrentWeather(latLong, weatherApiKey)
            _uiState.value = _uiState.value.copy(weather = weather)
        }
    }
}

class DetailViewModelFactory(
    private val repository: ParkRepository,
    private val parkCode: String,
    private val latLong: String,
    private val npsApiKey: String,
    private val trailApiKey: String,
    private val weatherApiKey: String,
    private val campgroundFields: String
) : ViewModelProvider.Factory {
    @Suppress("UNCHECKED_CAST")
    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        if (modelClass.isAssignableFrom(DetailViewModel::class.java)) {
            return DetailViewModel(
                repository,
                parkCode,
                latLong,
                npsApiKey,
                trailApiKey,
                weatherApiKey,
                campgroundFields
            ) as T
        }
        throw IllegalArgumentException("Unknown ViewModel class")
    }
}

